designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,284 @@
+
+## Backwards Compatibility Analysis
+
+This proposal contains two breaking changes:
+
+- If a schema declares `default` values, rule options that were `undefined` will become those defaults instead
+- Options will recursively default to `{}` based on schemas even if the schema has no `default` properties
+
+Those breaking changes will impact rules that rely on the exact length or existing of properties in `context.options`.
+Rules that wish to preserve the legacy options behavior can switch to referring to the new `context.optionsRaw` property.
+
+In the draft PR containing this change, only 3 test files -rule tests for `indent`, `max-len`, and `no-multiple-empty-lines`- failed during `npm run test`.
+All were fixed with a direct find-and-replace from `context.options` to `context.optionsRaw` in those rules' implementations.
+
+## Alternatives
+
+### Non-Recursive Object Defaults
+
+An alternate, lower-impact change would be to _not_ options to `{}` if there is no `default` in its schema.
+However:
+
+- Defaulting to `{}` allows rules logic to not need to `|| {}` or `?? {}` as much in source code.
+- Consider the case of an options schema declaring a `default` for a child property inside an object whose parent does not have a `default` introduces complexity:
+  - If the defaulting logic didn't recursively create default objects, it could be confusing why the `default` value wasn't being used.
+  - If the defaulting logic were to recursively create default objects only when a descendent property contains `default`, the inconsistency could also be confusing to users (and would be more code to implement in ESLint).
+
+This RFC proposes always defaulting objects to `{}` for consistency.
+
+### Prior Art
+
+The only standardized options application I could find was `@typescript-eslint/eslint-plugin`'s `defaultOptions` property on rules.
+It's visible in the plugin's [Custom Rules guide](https://typescript-eslint.io/developers/custom-rules#rulecreator), which suggests users wrap rule creation with a `createRule`:
+
+```ts
+import { ESLintUtils } from "@typescript-eslint/utils";
+
+const createRule = ESLintUtils.RuleCreator(
+  (name) => `https://example.com/rule/${name}`
+);
+
+export const rule = createRule({
+  create(context) {
+    /* ... */
+  },
+  defaultOptions: [
+    /* ... */
+  ],
+  meta: {
+    /* ... */
+  },
+  name: "...",
+});
+```
+
+`createRule` applies a _deep_ merge rather than a shallow `Object.assign`: https://github.com/typescript-eslint/typescript-eslint/blob/324c4d31e46cbc95e8eb67d71792de9f49e65606/packages/utils/src/eslint-utils/applyDefault.ts.
+This has the benefit of allowing user-provided options to "merge" into complex objects, rather than requiring rewriting larger objects.
+
+However, deep merges are a more nuanced behavior, and may surprise users who assume the more common shallow merging or a different strategy in edge cases such as arrays.
```

Comment:
It sounds like deep merging is the more powerful behavior, so is the concern over surprising the user enough to forgo that?

Reply:
imo deep merging is a code smell, and there simply does not exist a universally agreed-upon algorithm for how to do it, and thus it's always better to force users to do it manually rather than trying to guess at the right algorithm.

Reply:
TypeScript-ESLint's rule utilities have used a deep merge since its inception with no issue. There's a few dozen plugins and a few hundred rules with varying shapes of options that use it.

We haven't had reports of people being annoyed or upset with how it works.

If you know it's there and for some reason you don't want a deep merge either an opt out option can exist or you can just work around it by not not using it and manually merging the options.

I think there's two cases in TypeScript-ESLint that we opt out of our default utils - `naming-convention` and `ban-types` - the former because we wanted an "all-or-nothing" style of defaults; and the latter because we wanted to provide users the choice to "override and replace" or "augment and extend" the default options.

Your mileage may vary but the system has worked well for our corner of the ecosystem for quite a few years now.